File: src/main/java/no/novari/flyt/configuration/mapping/ObjectsFromCollectionMappingMappingService.java
package no.novari.flyt.configuration.mapping;

import no.novari.flyt.configuration.model.configuration.dtos.FromCollectionMappingDto;
import no.novari.flyt.configuration.model.configuration.dtos.ObjectMappingDto;
import no.novari.flyt.configuration.model.configuration.entities.collection.ObjectsFromCollectionMapping;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;


@Service
public class ObjectsFromCollectionMappingMappingService {

    private final InstanceCollectionReferencesMappingService instanceCollectionReferencesMappingService;
    private final ObjectMappingMappingService objectMappingMappingService;

    public ObjectsFromCollectionMappingMappingService(
            InstanceCollectionReferencesMappingService instanceCollectionReferencesMappingService,
            @Lazy ObjectMappingMappingService objectMappingMappingService
    ) {
        this.instanceCollectionReferencesMappingService = instanceCollectionReferencesMappingService;
        this.objectMappingMappingService = objectMappingMappingService;
    }

    public ObjectsFromCollectionMapping toEntity(
            FromCollectionMappingDto<ObjectMappingDto> objectsFromCollectionMappingDto
    ) {
        return ObjectsFromCollectionMapping
                .builder()
                .instanceCollectionReferencesOrdered(
                        instanceCollectionReferencesMappingService.toEntity(
                                objectsFromCollectionMappingDto.getInstanceCollectionReferencesOrdered()
                        )
                )
                .objectMapping(
                        objectMappingMappingService.toEntity(objectsFromCollectionMappingDto.getElementMapping())
                )
                .build();
    }

    public FromCollectionMappingDto<ObjectMappingDto> toDto(ObjectsFromCollectionMapping objectsFromCollectionMapping) {
        return FromCollectionMappingDto
                .<ObjectMappingDto>builder()
                .instanceCollectionReferencesOrdered(
                        instanceCollectionReferencesMappingService.toDto(
                                objectsFromCollectionMapping.getInstanceCollectionReferencesOrdered()
                        )
                )
                .elementMapping(objectMappingMappingService.toDto(objectsFromCollectionMapping.getObjectMapping()))
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/mapping/ObjectsFromCollectionMappingMappingService.kt
package no.novari.flyt.configuration.mapping

import no.novari.flyt.configuration.model.configuration.dtos.FromCollectionMappingDto
import no.novari.flyt.configuration.model.configuration.dtos.ObjectMappingDto
import no.novari.flyt.configuration.model.configuration.entities.collection.ObjectsFromCollectionMapping
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Service

@Service
class ObjectsFromCollectionMappingMappingService(
    private val instanceCollectionReferencesMappingService: InstanceCollectionReferencesMappingService,
    @param:Lazy private val objectMappingMappingService: ObjectMappingMappingService,
) {
    fun toEntity(
        objectsFromCollectionMappingDto: FromCollectionMappingDto<ObjectMappingDto>,
    ): ObjectsFromCollectionMapping =
        ObjectsFromCollectionMapping
            .builder()
            .instanceCollectionReferencesOrdered(
                instanceCollectionReferencesMappingService.toEntity(
                    objectsFromCollectionMappingDto.instanceCollectionReferencesOrdered,
                ),
            ).objectMapping(
                objectMappingMappingService.toEntity(
                    requireNotNull(objectsFromCollectionMappingDto.elementMapping),
                ),
            ).build()

    fun toDto(objectsFromCollectionMapping: ObjectsFromCollectionMapping): FromCollectionMappingDto<ObjectMappingDto> =
        FromCollectionMappingDto
            .builder<ObjectMappingDto>()
            .instanceCollectionReferencesOrdered(
                instanceCollectionReferencesMappingService.toDto(
                    objectsFromCollectionMapping.instanceCollectionReferencesOrdered,
                ),
            ).elementMapping(
                objectMappingMappingService.toDto(
                    requireNotNull(objectsFromCollectionMapping.objectMapping),
                ),
            ).build()
}
